### PR TITLE
Dyson orbital coefficients

### PR DESCRIPTION
**Pull request by tmhuysen**
**Short description**
Contains a function to calculate the coefficients of a Dyson orbital between two wave functions differing a single electron and which are expressed in the same basis. 

**Related issues**
Partially resolves issue #422

**Additional info**

**To-Do**
Q: Should we have reference data before proceeding? How can we test if the code is correct?

- [x] A: Perform a small calculation by hand, and use this as reference.

**Follow-up**

A follow up PR will contain a pybinding for quick use

### CODE REVIEW

File: include/Properties/properties.hpp [hunk 1]
```diff
@@ -20,7 +20,7 @@
 
 #include "Operator/SecondQuantized/SQOneElectronOperator.hpp"
 #include "RDM/OneRDM.hpp"
-
+#include "WaveFunction/WaveFunction.hpp"
```
Comment (lelemmen) [inline include/Properties/properties.hpp]: add newline

File: include/Properties/properties.hpp [hunk 2]
```diff
@@ -33,6 +33,12 @@ namespace GQCP {
...
 Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOperator<double>& dipole_operator, const OneRDM<double>& one_rdm);
 
-
+/**
+ *  @param wavefunction1        a wave function in a product Fock space  
+ *  @param wavefunction2        a wave function in a product Fock space containing one fewer electron and the same amount of orbitals that is expressed in the same basis
+ *  
+ *  @return a vector with the coefficients of a Dyson orbital derived from the difference between the two wave functions expressed in the basis of the wave functions
+ */
+VectorX<double> calculateDysonAmplitudes(const WaveFunction& wavefunction1, const WaveFunction& wavefunction2);
```
Comment (lelemmen) [inline include/Properties/properties.hpp]: add newline

File: include/Properties/properties.hpp [hunk 3]
```diff
@@ -33,6 +33,12 @@ namespace GQCP {
  */
 Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOperator<double>& dipole_operator, const OneRDM<double>& one_rdm);
 
-
+/**
```
Comment (lelemmen) [inline include/Properties/properties.hpp]:
```suggestion
/**
 *  Calculate the Dyson 'amplitudes' (the coefficients of a Dyson orbital) between two wave function expressed in the same spinor basis
```

File: include/Properties/properties.hpp [hunk 4]
```diff
@@ -33,6 +33,12 @@ namespace GQCP {
...
+ *  @param wavefunction1        a wave function in a product Fock space  
+ *  @param wavefunction2        a wave function in a product Fock space containing one fewer electron and the same amount of orbitals that is expressed in the same basis
+ *  
+ *  @return a vector with the coefficients of a Dyson orbital derived from the difference between the two wave functions expressed in the basis of the wave functions
```
Comment (lelemmen) [inline include/Properties/properties.hpp]:
```suggestion
 *  @return a vector with the Dyson orbital amplitudes
```

File: src/Properties/properties.cpp [hunk 5]
```diff
@@ -19,6 +19,7 @@
 
 #include "Properties/expectation_values.hpp"
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: remove newline

File: src/Properties/properties.cpp [hunk 6]
```diff
@@ -19,6 +19,7 @@
...
 
+#include "FockSpace/ProductFockSpace.hpp"
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: add newline

File: src/Properties/properties.cpp [hunk 7]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
 }
 
 
+/**
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: also update documentation here

File: src/Properties/properties.cpp [hunk 8]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+ *  @param wavefunction1        a wave function in a product Fock space  
+ *  @param wavefunction2        a wave function in a product Fock space containing one fewer electron and the same amount of orbitals that is expressed in the same basis
+ *  
+ *  @return a vector with the coefficients of a Dyson orbital derived from the difference between the two wave functions expressed in the basis of the wave functions
+ */
+VectorX<double> calculateDysonAmplitudes(const WaveFunction& wavefunction1, const WaveFunction& wavefunction2) {
+
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: Maybe add documentation in the likes of 'Check the arguments'

File: src/Properties/properties.cpp [hunk 9]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+    }
+
+    if (wavefunction2.get_fock_space().get_type() != FockSpaceType::ProductFockSpace) {
+        throw std::runtime_error("properties::calculateDysonOrbital(WaveFunction, WaveFunction): wavefunction2 is not in a product Fock space");
+    }
+
+    const auto fock_space1 = static_cast<const ProductFockSpace&>(wavefunction1.get_fock_space());
+    const auto fock_space2 = static_cast<const ProductFockSpace&>(wavefunction2.get_fock_space());
+
+    if (!((fock_space1.get_N_alpha() - fock_space2.get_N_alpha() == 0) && (fock_space1.get_N_beta() - fock_space2.get_N_beta() == 1))) {
```
Comment (lelemmen) [inline src/Properties/properties.cpp]:
```suggestion
    if ((fock_space1.get_N_alpha() - fock_space2.get_N_alpha() != 0) && (fock_space1.get_N_beta() - fock_space2.get_N_beta() != 1)) {
```

File: src/Properties/properties.cpp [hunk 10]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+        if (!((fock_space1.get_N_alpha() - fock_space2.get_N_alpha() == 1) && (fock_space1.get_N_beta() - fock_space2.get_N_beta() == 0))) {
```
Comment (lelemmen) [inline src/Properties/properties.cpp]:
```suggestion
        if ((fock_space1.get_N_alpha() - fock_space2.get_N_alpha() != 1) && (fock_space1.get_N_beta() - fock_space2.get_N_beta() != 0)) {
```

File: src/Properties/properties.cpp (hunk 9 above)
Comment (lelemmen) [inline src/Properties/properties.cpp]: Maybe move this before casting the Fock spaces

File: src/Properties/properties.cpp [hunk 11]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+            throw std::runtime_error("properties::calculateDysonOrbital(WaveFunction, WaveFunction): wavefunction2 is not expressed in a product Fock space with one fewer electron than wavefunction1");
+        }
+    } 
+
+    const auto fock_space_alpha1 = fock_space1.get_fock_space_alpha();
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: Can't we use `const auto&` in these cases?

Comment (lelemmen) [inline src/Properties/properties.cpp]: Add some documentation like 'Prepare some variables for the calculations'

File: src/Properties/properties.cpp [hunk 12]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+    const auto fock_space_alpha2 = fock_space2.get_fock_space_alpha();
+    const auto fock_space_beta1 = fock_space1.get_fock_space_beta();
+    const auto fock_space_beta2 = fock_space2.get_fock_space_beta();
+
+    const auto ci_coeff1 = wavefunction1.get_coefficients();
```
Comment (lelemmen) [inline src/Properties/properties.cpp]:
```suggestion
    const auto ci_coeffs1 = wavefunction1.get_coefficients();
```

File: src/Properties/properties.cpp [hunk 13]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+    const auto ci_coeff2 = wavefunction2.get_coefficients();
```
Comment (lelemmen) [inline src/Properties/properties.cpp]:
```suggestion
    const auto ci_coeffs2 = wavefunction2.get_coefficients();
```

File: src/Properties/properties.cpp [hunk 14]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+
+    VectorX<double> dyson_coeff = VectorX<double>::Zero(fock_space1.get_K());
+    size_t dim_alpha = fock_space_alpha1.get_dimension();
+    size_t dim_beta = fock_space_beta1.get_dimension();
+
+    // Given the one electron difference requirement, one of the spin Fock spaces will be identical for both wave functions, identifying that spin Fock space allows for a simple algorithm
+
+    // Beta electron differs by one
+    if ((fock_space1.get_N_alpha() - fock_space2.get_N_alpha() == 0) && (fock_space1.get_N_beta() - fock_space2.get_N_beta() == 1)) {
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: Explain the main parts of your algorithm in the comments

File: src/Properties/properties.cpp [hunk 15]
```diff
@@ -38,4 +39,114 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+            }
+
+
+            if (Ib < dim_beta - 1) {  // prevent last permutation to occur
+                fock_space_beta1.setNextONV(onv_beta);
+            }
+        }  // beta address (Ib) loop
+    }
+
+    // Alpha electrons differs by one
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: idem: explain

Comment (lelemmen) [reply src/Properties/properties.cpp]: This seems like a copy-paste: is there a way you can remove some duplicate codes?

File: tests/Properties/properties_test.cpp [hunk 16]
```diff
@@ -24,6 +24,9 @@
 #include "Properties/properties.hpp"
 #include "RHF/DIISRHFSCFSolver.hpp"
 #include "RHF/PlainRHFSCFSolver.hpp"
+#include "HamiltonianBuilder/FCI.hpp"
```
Comment (lelemmen) [inline tests/Properties/properties_test.cpp]: put in alphabetical order

File: tests/Properties/properties_test.cpp [hunk 17]
```diff
@@ -101,3 +104,28 @@ BOOST_AUTO_TEST_CASE ( dipole_N2_STO_3G ) {
     GQCP::Vector<double, 3> total_dipole_moment = GQCP::Operator::NuclearDipole(N2).value() + GQCP::calculateElectronicDipoleMoment(dipole_op, D);
     BOOST_CHECK(std::abs(total_dipole_moment.norm() - (0.0)) < 1.0e-08);
 }
+
+
+/**
+ *  Test the dyson amplitude algorithm against manually calculated amplitudes for two (normalized) toy vectors
```
Comment (lelemmen) [inline tests/Properties/properties_test.cpp]:
```suggestion
 *  Test the Dyson amplitude algorithm against manually calculated amplitudes for two (normalized) toy wave functions
```

File: tests/Properties/properties_test.cpp [hunk 18]
```diff
@@ -101,3 +104,28 @@ BOOST_AUTO_TEST_CASE ( dipole_N2_STO_3G ) {
...
+ */
+BOOST_AUTO_TEST_CASE ( dyson_amplitudes) {
+
+    GQCP::VectorX<double> reference_amplitudes = GQCP::VectorX<double>::Zero(2); reference_amplitudes << 0.46752445464799997, 0.8415438533449999;
```
Comment (lelemmen) [inline tests/Properties/properties_test.cpp]: Split this over two lines

File: tests/Properties/properties_test.cpp [hunk 19]
```diff
@@ -101,3 +104,28 @@ BOOST_AUTO_TEST_CASE ( dipole_N2_STO_3G ) {
     GQCP::Vector<double, 3> total_dipole_moment = GQCP::Operator::NuclearDipole(N2).value() + GQCP::calculateElectronicDipoleMoment(dipole_op, D);
     BOOST_CHECK(std::abs(total_dipole_moment.norm() - (0.0)) < 1.0e-08);
 }
+
+
+/**
+ *  Test the dyson amplitude algorithm against manually calculated amplitudes for two (normalized) toy vectors
+ */
+BOOST_AUTO_TEST_CASE ( dyson_amplitudes) {
```
Comment (lelemmen) [inline tests/Properties/properties_test.cpp]: Put as many `const`s as possible

File: tests/Properties/properties_test.cpp [hunk 20]
```diff
@@ -101,3 +104,28 @@ BOOST_AUTO_TEST_CASE ( dipole_N2_STO_3G ) {
...
+
+    GQCP::VectorX<double> reference_amplitudes = GQCP::VectorX<double>::Zero(2); reference_amplitudes << 0.46752445464799997, 0.8415438533449999;
+
+    size_t K = 2;
+    size_t N = 2;
+ 
+    GQCP::ProductFockSpace fock_space1 (K, N/2, N/2);
+    GQCP::ProductFockSpace fock_space2 (K, N/2, N/2-1);
+  
+    GQCP::VectorX<double> vec1 = GQCP::VectorX<double>::Zero(4); vec1 << 0.182574, 0.365148, 0.547723, 0.730297;
```
Comment (lelemmen) [inline tests/Properties/properties_test.cpp]: Split over two lines

File: tests/Properties/properties_test.cpp [hunk 21]
```diff
@@ -101,3 +104,28 @@ BOOST_AUTO_TEST_CASE ( dipole_N2_STO_3G ) {
...
+    GQCP::VectorX<double> vec2 = GQCP::VectorX<double>::Zero(2); vec2 << 0.640184, 0.768221;
```
Comment (lelemmen) [inline tests/Properties/properties_test.cpp]: Split over two lines

File: tests/Properties/properties_test.cpp [hunk 22]
```diff
@@ -101,3 +104,28 @@ BOOST_AUTO_TEST_CASE ( dipole_N2_STO_3G ) {
...
+    
+    const auto wavefunction1 = GQCP::WaveFunction(fock_space1, vec1);
+    const auto wavefunction2 = GQCP::WaveFunction(fock_space2, vec2);
+
+    const auto dyson_coefficients = GQCP::calculateDysonAmplitudes(wavefunction1, wavefunction2);
+
+    BOOST_CHECK(dyson_coefficients.isApprox(reference_amplitudes, 10e-6));
```
Comment (lelemmen) [inline tests/Properties/properties_test.cpp]:
```suggestion
    BOOST_CHECK(dyson_coefficients.isApprox(reference_amplitudes, 1.0e-6));
```

File: src/Properties/properties.cpp (hunk 9 above)
Comment (tmhuysen) [reply src/Properties/properties.cpp]: Can't access these variables before casting

File: src/Properties/properties.cpp (hunk 15 above)
Comment (tmhuysen) [reply src/Properties/properties.cpp]: You are correct, I have modified this, however it does look a tad more complicated

**Comment by lelemmen** [comment]
@guacke @tmhuysen What do you think about renaming the function to `calculateDysonOrbitalCoefficients` instead of the `calculateDysonAmplitudes`? The Dyson orbital amplitudes are the separate contributions for each of the Dyson orbital coefficients, so I think that should be reflected in the naming. I'm sorry for the confusion.

File: src/Properties/properties.cpp [hunk 23]
```diff
@@ -38,4 +39,101 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+        if ((fock_space1.get_N_alpha() - fock_space2.get_N_alpha() != 1) && (fock_space1.get_N_beta() - fock_space2.get_N_beta() != 0)) {
+            throw std::runtime_error("properties::calculateDysonOrbital(WaveFunction, WaveFunction): wavefunction2 is not expressed in a product Fock space with one fewer electron than wavefunction1");
+        }
+    } 
+
+    // Initialize environment variables
+
+    // The 'passive' spin Fock spaces are the Fock spaces that are equal for both wave functions
+    // The 'target' spin Fock spaces have an electron difference of one
+    // We initialize the variables for the case in which they differ in one beta electron, if this isn't the case, we will update it later    auto passive_fock_space1 = fock_space1.get_fock_space_alpha();
```
Comment (lelemmen) [inline src/Properties/properties.cpp]:
```suggestion
    // We initialize the variables for the case in which they differ in one beta electron, if this isn't the case, we will update it later
```

File: src/Properties/properties.cpp [hunk 24]
```diff
@@ -38,4 +39,101 @@ Vector<double, 3> calculateElectronicDipoleMoment(const VectorSQOneElectronOpera
...
+    // Since we want to calculate the overlap between two wave functions, the ONVs should have an equal number of electrons
+    // We therefore iterate over the ONVs of the 'target' Fock space, which all have an electron more, and annihilate in one of the orbitals (let the index of that orbital be called 'p')
+    // By calculating the overlap in the (N-1)-Fock space, we can calculate the contributions to the  'p'-th coefficient (i.e. the Dyson amplitude) of the Dyson orbital
+    ONV onv = target_fock_space1.makeONV(0);
+
+    for (size_t It = 0; It < target_fock_space1.get_dimension(); It++) {  // It loops over addresses of the target Fock space
+        int sign = -1;  // total phase factor of all the annihilations that have occurred
+        for (size_t e = 0; e < target_fock_space1.get_N(); e++) {  // loop over electrons in the ONV
+            
+            sign *= -1;
```
Comment (lelemmen) [inline src/Properties/properties.cpp]: Could you move lines 112 and 113 below the comment that starts on 115? In this way, the code is more structured.

**Pull request merged**